### PR TITLE
Add check for ROS2 distribution in CMakeLists.txt

### PR DESCRIPTION
**Pull request by arturkamieniecki**
I've modified CMakeLists.txt to check if the initial configuration was run under the same ROS 2 distribution as the one sourced during project setup and build. I've changed CMakeLists.txt such that the ROS 2 distribution will be cached into CMakeCache.txt during the project's configuration. A check during the configuration process prevents users from building when their sourced distribution is different. I've also added the same check during the build process.

Removing the CMakeCache.txt, as proposed by me (```Remove CMakeCache.txt and reconfigure project.```), does not cause rebuilding of the O3DE Engine.

This addresses #135.

### CODE REVIEW

**Review by lawsonamzn — COMMENTED** [review]
looks like it will work - is it necessary to remove the cache, though?
Does it cache stuff that it doesn't recompute even if you do a reconfigure? 

To be more specific, do we have to force a cache deletion here, or can we force an auto-reconfigure?

**Comment by lumberyard-employee-dm** [comment]
> looks like it will work - is it necessary to remove the cache, though? Does it cache stuff that it doesn't recompute even if you do a reconfigure?
> 
> To be more specific, do we have to force a cache deletion here, or can we force an auto-reconfigure, since a lot of stuff is checked in the configuration step but a problem is that env vars changing doesn't cause a reconfigure...

An alternative is using the `-UROS_DISTRO_TYPE` option to undefine the `ROS_DISTRO_TYPE` cache variable to force a new check of the ROS2 distro

**Comment by lawsonamzn** [comment]
> 

I'm thinking more along the lines of what inside the cache is never checked again even on reconfigure.  An example is things like when find_program or find_library is used, unless those vars are cleared, or FORCE is used, cmake tends to cache them.  I ran into this quite a bit when dealing with the 3p system.

File: Gems/ROS2/Code/checkROS2Distribution.cmake [hunk 1]
```diff
@@ -0,0 +1,3 @@
+if (NOT "$ENV{ROS_DISTRO}" STREQUAL "${ROS_DISTRO}")
```
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/checkROS2Distribution.cmake:1]: This file is missing a copyright and a newline at the end as well.

**Comment by lumberyard-employee-dm** [comment]
> > 
> 
> I'm thinking more along the lines of what inside the cache is never checked again even on reconfigure. An example is things like when find_program or find_library is used, unless those vars are cleared, or FORCE is used, cmake tends to cache them. I ran into this quite a bit when dealing with the 3p system. To actually force those you'd have to also have to undef (in the cache) the vars that are caching those program locations.

How about checking if the value in the Environment variable is different than the one in the cache if the cache variable is set?
If so force update the cache with the new value.

**Comment by lawsonamzn** [comment]
> > > 
> > 
> > 
> > I'm thinking more along the lines of what inside the cache is never checked again even on reconfigure. An example is things like when find_program or find_library is used, unless those vars are cleared, or FORCE is used, cmake tends to cache them. I ran into this quite a bit when dealing with the 3p system. To actually force those you'd have to also have to undef (in the cache) the vars that are caching those program locations.
> 
> How about checking if the value in the Environment variable is different than the one in the cache if the cache variable is set? If so force update the cache with the new value.
> 
> That can be an option instead of a `message(FATAL_ERROR)`

I think the one where it fails the build due to an error in the **build step** is valid, since CMake doesn't actually check the value of Env Vars by default during build if nothing has changed in its dependency tree.    

So maybe the flow would be

If you configure, it wipes the vars and configures with the new sources / env vars.
If you build _without configuring_ and it skips the confiugure step, the build step detects the mismatch, and forces a failure but also clears the cache variables or touches a file or something that wil cause the next attempt to auto-reconfigure?

**Comment by lumberyard-employee-dm** [comment]
> > > > 
> > > 
> > > 
> > > I'm thinking more along the lines of what inside the cache is never checked again even on reconfigure. An example is things like when find_program or find_library is used, unless those vars are cleared, or FORCE is used, cmake tends to cache them. I ran into this quite a bit when dealing with the 3p system. To actually force those you'd have to also have to undef (in the cache) the vars that are caching those program locations.
> > 
> > 
> > How about checking if the value in the Environment variable is different than the one in the cache if the cache variable is set? If so force update the cache with the new value.
> > That can be an option instead of a `message(FATAL_ERROR)`
> 
> I think the one where it fails the build due to an error in the **build step** is valid, since CMake doesn't actually check the value of Env Vars by default during build if nothing has changed in its dependency tree.
> 
> So maybe the flow would be
> 
> If you configure, it wipes the vars and configures with the new sources / env vars. If you build _without configuring_ and it skips the configure step, the build step detects the mismatch, and forces a failure but also clears the cache variables or touches a file or something that will cause the next attempt to auto-reconfigure?

Ok, that can work 95% of the way. The one thing that can't be done is to clear the cache from the build step since [checkROS2Distribution.cmake](https://github.com/o3de/o3de-extras/pull/345/files#diff-6b7b5fdec66b3029927dc7ddfeaa4d79a15ff3494a6a59258954820f18c9aa50) custom target doesn't have access to the cmake binary directory cache values since it is an external command.

File: Gems/ROS2/Code/checkROS2Distribution.cmake (hunk 1 above)
Comment (michalpelka) [inline Gems/ROS2/Code/checkROS2Distribution.cmake:1]: Do we need the same code in two places?

File: Gems/ROS2/Code/CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,23 @@ if (NOT ROS2_FOUND)
...
 message(DEBUG "Building ${gem_name} Gem with ros2 $ENV{ROS_DISTRO}")
 
+# Check if ROS 2 distribution is cached
+get_property(ROS_DISTRO_TYPE CACHE ROS_DISTRO PROPERTY TYPE)
+
+# Perform checks with cached ROS 2 distribution and sourced distribution. Save the distribution into cache if it was not cached before or if they do not match end with error.
+if (NOT ROS_DISTRO_TYPE)
+    set(ROS_DISTRO $ENV{ROS_DISTRO} CACHE STRING "ROS 2 distribution of current configuration" FORCE)
+elseif(NOT "${ROS_DISTRO}" STREQUAL "$ENV{ROS_DISTRO}")
+    message(FATAL_ERROR "ROS 2 distribution does not match. Configuration created for: ${ROS_DISTRO}, sourced distribution: $ENV{ROS_DISTRO}. Remove CMakeCache.txt and reconfigure project.")
```
[the remark this answers is not in the document]
Comment (lawsonamzn) [reply Gems/ROS2/Code/CMakeLists.txt:29]: Its the nuclear option but if it works and solves the problem, then its a lot easier to say "Configuration has been automatically reset, please try again" than to tell them to dig in their file system. :)

**Review by lawsonamzn — APPROVED** [review]
Nice.  Its a brute force method but from the user's point of view they will get the error, and then if they try again, cmake will reconfigure and retry automatically...

**Review by spham-amzn — APPROVED** [review]

**Review by michalpelka — APPROVED** [review]
:boom:

**Review by adamdbrw — APPROVED** [review]
Nice work!

**Comment by lawsonamzn** [comment]
I did a clean build, it looks like it passed.

**Pull request merged**